workers/unity/Packages/io.improbable.gdk.core/Collections/EntityRangeCollection.cs

```diff
@@ -0,0 +1,184 @@
+                else
+                {
+                    var splitRange = new EntityIdRange(firstElement.FirstEntityId, count);
+                    collection.Add(splitRange);
+                    firstElement.FirstEntityId = new EntityId(firstElement.FirstEntityId.Id + count);
+                    firstElement.Count -= count;
+                    break;
+                }
+            }
+
+            Count -= count;
+            version++;
+
+            return collection;
+        }
+
+        internal void Add(EntityIdRange range)
+        {
+            if (!range.FirstEntityId.IsValid())
+            {
+                throw new ArgumentException("Invalid EntityId in range", nameof(range));
+            }
+
+            if (range.Count == 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(range), "EntityId range cannot be 0 in length.");
+            }
+
+            if (firstElement.Count == 0)
+            {
+                firstElement = range;
+            }
+            else
+            {
+                queue.Enqueue(range);
+            }
+
+            Count += range.Count;
+            version++;
+        }
+
+        public void Clear()
+        {
+            firstElement = new EntityIdRange();
+            queue.Clear();
+            Count = 0;
+            version++;
+        }
+
+        internal struct EntityIdRange
+        {
+            public EntityId FirstEntityId;
+            public uint Count;
+
+            public EntityIdRange(EntityId firstEntityId, uint count)
+            {
+                Count = count;
+                FirstEntityId = firstEntityId;
+            }
+        }
+
+        public Enumerator GetEnumerator()
+        {
+            return new Enumerator(this);
+        }
+
+        IEnumerator<EntityId> IEnumerable<EntityId>.GetEnumerator()
+        {
+            return new Enumerator(this);
+        }
+
+        IEnumerator IEnumerable.GetEnumerator()
+        {
+            return GetEnumerator();
+        }
+
+        public struct Enumerator : IEnumerator<EntityId>
+        {
+            private EntityRangeCollection collection;
+
+            private EntityIdRange activeRange;
+            private Queue<EntityIdRange>.Enumerator enumerator;
+
+            private int index;
+            private int version;
+            private EntityId current;
+
+            public Enumerator(EntityRangeCollection collection)
+            {
+                this.collection = collection;
+                activeRange = collection.firstElement;
+                enumerator = collection.queue.GetEnumerator();
+
+                index = 0;
+                version = collection.version;
+                current = default;
+            }
+
+            public bool MoveNext()
+            {
+                if (version != collection.version)
+                {
+                    throw new InvalidOperationException("Cannot iterate over collection which has been altered.");
```

Comment:
Suggest coping the wording of the standard exceptions for these things if this isn't it already

Reply:
`Collection was modified; enumeration operation may not execute`